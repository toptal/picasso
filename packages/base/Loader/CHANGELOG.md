# @toptal/picasso-loader

## 3.0.0

### Major Changes

- [#4500](https://github.com/toptal/picasso/pull/4500) [`3ed8c02`](https://github.com/toptal/picasso/commit/3ed8c0271982a82dd9cdc6b967c63656afd3654f) Thanks [@ruslan-sed](https://github.com/ruslan-sed)!
- update version of `@toptal/picasso-tailwind-merge` peer dependency

## 2.0.0

### Major Changes

- [#4345](https://github.com/toptal/picasso/pull/4345) [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685) Thanks [@augustobmoura](https://github.com/augustobmoura)!
- add @toptal/picasso-tailwind-merge as peer dependency, it now needs to be provided by the project consuming the package

- [#4345](https://github.com/toptal/picasso/pull/4345) [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685) Thanks [@augustobmoura](https://github.com/augustobmoura)!

### Loader

- migrated away from MUI to pure Tailwind CSS

## 1.0.3

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-utils@1.0.3

## 1.0.2

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-utils@1.0.2

## 1.0.1

### Patch Changes

- [#4164](https://github.com/toptal/picasso/pull/4164) [`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30) Thanks [@mkrl](https://github.com/mkrl)!
- migrate to Picasso by Parts
  - picasso is now distributed as a set of independent packages with the main package `@toptal/picasso` now being a collection of re-exported packages
- Updated dependencies [[`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30)]:
  - @toptal/picasso-utils@1.0.1
