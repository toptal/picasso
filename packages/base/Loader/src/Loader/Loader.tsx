import type { ReactNode, HTMLAttributes } from 'react'
import React, { forwardRef } from 'react'
import type { BaseProps, SizeType } from '@toptal/picasso-shared'
import { useAppConfig } from '@toptal/picasso-provider'
import { twMerge, twJoin } from '@toptal/picasso-tailwind-merge'

const DEFAULT_PROGRESS = 35
const VIEWBOX_SIZE = 44
const THICKNESS = 3.6

enum SIZES {
  small = 16,
  medium = 32,
  large = 64,
}

type VariantType = 'blue' | 'inherit'

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** Text content for the `Loader` */
  children?: ReactNode
  /** Shows loader as part of other inline elements such as text */
  inline?: boolean
  /** Size of the `Loader` */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** Set the value if want to have static loader with the value specified */
  value?: number
  /** Set this value if you want loader to inherit color of the parent, primary by default */
  variant?: VariantType
}

export const Loader = forwardRef<HTMLDivElement, Props>(function Loader(
  props,
  ref
) {
  const {
    children,
    size = 'medium',
    inline = false,
    className,
    value,
    variant = 'blue',
    style,
    ...rest
  } = props

  const { disableTransitions } = useAppConfig()

  const progress = disableTransitions ? DEFAULT_PROGRESS : value
  const animated = !(disableTransitions || Boolean(progress))

  const circumference = 2 * Math.PI * ((VIEWBOX_SIZE - THICKNESS) / 2)

  return (
    <div
      {...rest}
      ref={ref}
      className={twMerge(
        'text-lg flex flex-col items-center',
        inline ? 'inline-flex' : '',
        className
      )}
    >
      <div
        className={twJoin(
          variant === 'blue' ? 'text-blue-500' : 'text-inherit',
          animated
            ? 'animate-[spin_1400ms_linear_infinite]'
            : 'transform -rotate-90'
        )}
        style={{ width: SIZES[size], height: SIZES[size], ...style }}
        role='progressbar'
        aria-valuenow={progress ? Math.round(progress) : undefined}
      >
        <svg
          className='block'
          viewBox={`${VIEWBOX_SIZE / 2} ${
            VIEWBOX_SIZE / 2
          } ${VIEWBOX_SIZE} ${VIEWBOX_SIZE}`}
        >
          <circle
            className={twJoin(
              animated
                ? 'animate-[stroke-dash_1.4s_ease-in-out_infinite]'
                : 'transition-all',
              'stroke-[currentColor]'
            )}
            cx={VIEWBOX_SIZE}
            cy={VIEWBOX_SIZE}
            r={(VIEWBOX_SIZE - THICKNESS) / 2}
            fill='none'
            strokeDasharray='80px, 200px'
            strokeDashoffset='0px'
            strokeWidth={THICKNESS}
            style={
              progress
                ? {
                    strokeDashoffset: `${(
                      ((100 - progress) / 100) *
                      circumference
                    ).toFixed(3)}px`,
                    strokeDasharray: circumference.toFixed(3),
                  }
                : undefined
            }
          />
        </svg>
      </div>

      {children && <div className='mt-4'>{children}</div>}
    </div>
  )
})

Loader.defaultProps = {
  inline: false,
  size: 'medium',
  variant: 'blue',
}

Loader.displayName = 'Loader'

export default Loader
