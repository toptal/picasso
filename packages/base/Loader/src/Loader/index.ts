import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Loader'

export { default as Loader } from './Loader'
export type LoaderProps = OmitInternalProps<Props>
