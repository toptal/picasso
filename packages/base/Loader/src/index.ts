export * from './Loader'
export * from './CircularProgress'