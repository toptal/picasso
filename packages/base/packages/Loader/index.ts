/* eslint-disable import/no-extraneous-dependencies */
import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Loader'

export { default } from './Loader'
export type LoaderProps = OmitInternalProps<Props>
