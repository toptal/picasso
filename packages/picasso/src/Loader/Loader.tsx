import React, { ReactNode, forwardRef, HTMLAttributes } from 'react'
import { makeStyles, Theme } from '@material-ui/core/styles'
import capitalize from '@material-ui/core/utils/capitalize'
import cx from 'classnames'
import { BaseProps, SizeType } from '@toptal/picasso-shared'
import { useAppConfig } from '@toptal/picasso-provider'

import CircularProgress from '../CircularProgress'
import styles from './styles'

const DEFAULT_PROGRESS = 35

enum SIZES {
  small = 16,
  medium = 32,
  large = 64,
}

type VariantType = 'blue' | 'inherit'

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** Text content for the `Loader` */
  children?: ReactNode
  /** Shows loader as part of other inline elements such as text */
  inline?: boolean
  /** Size of the `Loader` */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** Set the value if want to have static loader with the value specified */
  value?: number
  /** Set this value if you want loader to inherit color of the parent, primary by default */
  variant?: VariantType
}

const useStyles = makeStyles<Theme>(styles, { name: 'PicassoLoader' })

export const Loader = forwardRef<HTMLDivElement, Props>(function Loader(
  props,
  ref
) {
  const {
    children,
    size = 'medium',
    inline = false,
    className,
    value,
    variant = 'blue',
    ...rest
  } = props

  const classes = useStyles()
  const { disableTransitions } = useAppConfig()

  const progress = disableTransitions ? DEFAULT_PROGRESS : value
  const progressVariant =
    disableTransitions || value ? 'static' : 'indeterminate'

  return (
    <div
      {...rest}
      ref={ref}
      className={cx(classes.root, className, {
        [classes.inline]: inline,
      })}
    >
      <CircularProgress
        classes={{
          root: classes[`spinner${capitalize(variant)}`],
        }}
        size={SIZES[size]}
        value={progress}
        variant={progressVariant}
      />

      {children && <div className={classes.label}>{children}</div>}
    </div>
  )
})

Loader.defaultProps = {
  inline: false,
  size: 'medium',
  variant: 'blue',
}

Loader.displayName = 'Loader'

export default Loader
