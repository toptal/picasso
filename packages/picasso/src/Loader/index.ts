import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Loader'

export { default } from './Loader'
export type LoaderProps = OmitInternalProps<Props>
