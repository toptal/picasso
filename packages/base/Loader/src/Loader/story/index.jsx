import { Loader } from '../Loader'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Loader',
  `Loaders indicate that an action is underway and that the user must wait to proceed until it is finished.
  
  ${PicassoBook.createSourceLink(__filename)}
  `
)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: Loader, name: 'Loader' })

page
  .createChapter()
  .addExample('Loader/story/Default.example.tsx', 'Default', 'base/Loader')
  .addExample('Loader/story/WithLabel.example.tsx', 'With label', 'base/Loader')
  .addExample(
    'Loader/story/Inline.example.tsx',
    'With inline content',
    'base/Loader'
  )
  .addExample('Loader/story/Sizes.example.tsx', 'Sizes', 'base/Loader')
  .addExample('Loader/story/Variants.example.tsx', 'Variants', 'base/Loader')
  .addExample(
    'Loader/story/ControlledValue.example.tsx',
    {
      title: 'Controlled value',
      description: 'Loader with determined or static values',
    },
    'base/Loader'
  )
