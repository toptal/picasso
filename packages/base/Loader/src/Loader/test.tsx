import type { ReactNode } from 'react'
import React from 'react'
import { render } from '@testing-library/react'
import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Loader'
import Loader from './Loader'

const renderLoader = (children: ReactNode, props: OmitInternalProps<Props>) => {
  const { inline, size } = props

  return render(
    <Loader inline={inline} size={size}>
      {children}
    </Loader>
  )
}

describe('Loader', () => {
  it('props combo', () => {
    const { container } = renderLoader('Testing', {
      inline: true,
      size: 'large',
    })

    expect(container).toMatchSnapshot()
  })
})
