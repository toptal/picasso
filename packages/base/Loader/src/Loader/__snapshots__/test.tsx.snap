// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader props combo 1`] = `
<div>
  <div
    class="text-lg flex-col items-center inline-flex"
  >
    <div
      class="text-blue animate-[spin_1400ms_linear_infinite]"
      role="progressbar"
      style="width: 64px; height: 64px;"
    >
      <svg
        class="block"
        viewBox="22 22 44 44"
      >
        <circle
          class="animate-[stroke-dash_1.4s_ease-in stroke-[currentColor]"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-dasharray="80px, 200px"
          stroke-dashoffset="0px"
          stroke-width="3.6"
        />
      </svg>
    </div>
    <div
      class="mt-4"
    >
      Testing
    </div>
  </div>
</div>
`;
